src/pyhf/workspace.py

```diff
@@ -16,11 +18,262 @@
 log = logging.getLogger(__name__)
 
 
+def _join_items(join, left_items, right_items, key='name'):
+    """
+    Join two lists of dictionaries along the given key.
+
+    This is meant to be as generic as possible for any pairs of lists of dictionaries for many join operations.
+
+    Args:
+        join (str): The join operation to apply. See ~pyhf.workspace.Workspace for valid join operations.
+        left_items (list): A list of dictionaries to join on the left
+        right_items (list): A list of dictionaries to join on the right
+
+    Returns:
+        :obj:`list`: A joined list of dictionaries.
+
+    """
+    if join == 'right outer':
+        left_items, right_items = right_items, left_items
```

Comment:
maybe better to call one `primary` and the other `secondary` and assign `left` `right` according to the passed preference order?